### PR TITLE
feat(qiankun): 支持 qiankun2.0 特性

### LINKED ISSUE #238: package.json依赖中的qiankun plugin版本问题
image
package.json依赖中的qiankun plugin版本回退至1.xx了，出现功能问题

### PR DESCRIPTION
**Pull request by kuitos**
resolve https://github.com/umijs/plugins/issues/64
resolve #238 
resolve https://github.com/umijs/umi/issues/4624
resolve https://github.com/umijs/umi/issues/4323
resolve https://github.com/umijs/umi/issues/4282
resolve https://github.com/umijs/umi/issues/4010
resolve https://github.com/umijs/umi/issues/3919

close https://github.com/umijs/plugins/pull/191
close https://github.com/umijs/umi/issues/4843
close https://github.com/umijs/umi/issues/4280

文档 https://github.com/umijs/umi/pull/4427

---
This change is [Reviewable](https://reviewable.io/reviews/umijs/plugins/148)

### CODE REVIEW

File: packages/plugin-qiankun/src/master/modifyRoutes.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+          if (route.routes?.length) {
+            throw new Error(
+              '[@umijs/plugin-qiankun]: You can not attach micro app to a route who has children!',
+            );
+          }
+
+          const { settings = {} } = route;
+          route.exact = false;
+          route.component = `({match}) => {
+            const MicroApp = require('umi').MicroApp as any;
```
Comment (sorrycc) [inline packages/plugin-qiankun/src/master/modifyRoutes.ts:101]: 这个会导致 umi 全量编译，没有 tree-shaking。

File: packages/plugin-qiankun/src/master/modifyRoutes.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+function modifyRoutesWithAttachMode(api: IApi) {
+  const { history } = api.service.userConfig;
+  const masterHistoryType = history?.type || defaultHistoryType;
+
+  const patchRoutes = (routes: IRoute[]) => {
+    if (routes.length) {
+      routes.forEach(route => {
+        const { microApp } = route;
```
Comment (sorrycc) [inline packages/plugin-qiankun/src/master/modifyRoutes.ts:90]: 不要写死了。

**Comment by sorrycc** [comment]
冲突修下，其他没问题。

**Comment by sorrycc** [comment]
windows 下 ci 没过，排查下。

**Comment by 957739315** [comment]
请问这个冲突问题，是已经修复好了吗?是修复到@umijs/plugin-qiankun哪个版本下呢？[history和single-spa冲突，单一回调会重复执行两次，导致umi和qiankun一起使用时行为异常 umijs/umi#3919](https://github.com/umijs/umi/issues/3919)

**Pull request merged**